modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/PlanningContext.java

```diff
@@ -210,6 +232,75 @@ public RexBuilder rexBuilder() {
         return unwrap(BaseQueryContext.class).rexBuilder();
     }
 
+    /**
+     * Stores skipped hint and the reason. Also, logs the issue.
+     */
+    public void skippedHint(RelNode relNode, RelHint hint, String reason) {
+        skippedHints.compute(new IgniteBiTuple<>(hint, relNode), (k, sh) -> {
```

Comment:
I don't like the idea of storing problematic hints in planning context. Planner can create copies of relNode (for example, when we expand or to union with two table scans) with the same hints and all these duplicated hints will be stored here. Currentrly, we use skippedHint method only once, to check that we don't have conflicting NO_INDEX/FORCE_INDEX hints. Can we instead do this check on validation before planning? In this case patch will be simplifeed and we can revert changes to `PlanningContext`, `Commons` and some other related classes.

Reply:
I don't think that the validation phase is actually handy for hints. On this phase, hints are not fully parsed, have no parsed params containing. Aren't filtered with the hint strategies, are not correctly set for suitable nodes. For instance, in `validateJoin(SqlJoin join, SqlValidatorScope scope)` `join` has no hints which might be actually set in related `select`. You need to find, filter, parse them manually in the related scope. This job is done further by Calcite. In a rule, you get correctly assigned hint for current node with parsed params. Here, in single place, we can check index name, table names etc. and apply (or not) hint. No code split by two or several places. Looks much simpler.